packages/foundry/contracts/hooks/rebalancer/Rebalancer.sol

```diff
@@ -0,0 +1,633 @@
+     * @dev The param indicates various places the error has accoured
+     * example: "PW_N_PT" is Pool weights do not match pool tokens
+     */
+    error InvalidArrayLenghts(string message);
+
+    /**
+     * @notice Hooks functions called from an external router.
+     * @dev This contract inherits both `MinimalRouter` and `BaseHooks`, and functions as is its own router.
+     * @param router The address of the Router
+     */
+    error CannotUseExternalRouter(address router);
+
+    constructor(
+        IVault vault,
+        IPermit2 permi2,
+        IWETH weth,
+        address _weightedPoolFactory,
+        address _oracle
+    ) MinimalRouterWithSwap(vault, weth, permi2) Ownable(msg.sender) {
+        weightedPoolFactory = _weightedPoolFactory;
+        oracle = _oracle;
+    }
+
+    modifier onlySelfRouter(address router) {
+        _ensureSelfRouter(router);
+        _;
+    }
+    ///@dev prevent race conditions when rebalancing
+    bool private isRebalancing;
+
+    modifier nonReentrantRebalance() {
+        require(!isRebalancing, "Rebalance in progress");
+        isRebalancing = true;
+        _;
+        isRebalancing = false;
+    }
+
+    /***************************************************************************
+                                  Router Functions
+    ***************************************************************************/
+    function addLiquidityProportional(
+        address pool,
+        uint256[] memory maxAmountsIn,
+        uint256 exactBptAmountOut,
+        bool wethIsEth,
+        bytes memory userData
+    ) external payable saveSender returns (uint256[] memory amountsIn) {
+        // Do addLiquidity operation - BPT is minted to this contract.
+        amountsIn = _addLiquidityProportional(
+            pool,
+            msg.sender,
+            address(this),
+            maxAmountsIn,
+            exactBptAmountOut,
+            wethIsEth,
+            userData
+        );
+        address[] storage lps = liquidityProviders[pool];
+        lps.push(msg.sender);
+        uint256[] storage amounts = amountTokens[pool][msg.sender];
+        for (uint256 i = 0; i < amounts.length; i++) {
+            amounts[i] += amountsIn[i];
+        }
+        isStillLp[pool][msg.sender] = true;
+        emit LiquidtyProviderAdded(pool, msg.sender);
+    }
+
+    function removeLiquidityProportional(
+        uint256 tokenId,
+        address pool,
+        uint256 exactBptAmountIn,
+        uint256[] memory minAmountsOut,
+        bool wethIsEth
+    ) external payable saveSender returns (uint256[] memory amountsOut) {
+        // Do removeLiquidity operation - tokens sent to msg.sender.
+        amountsOut = _removeLiquidityProportional(
+            pool,
+            msg.sender,
+            msg.sender,
+            exactBptAmountIn,
+            minAmountsOut,
+            wethIsEth,
+            ""
+        );
+        uint256[] memory lpBptAmount = amountTokens[pool][msg.sender];
```

Comment:
The name is confusing here; you're tracking the portion of this pool's token balances for this sender, not the BPT

Reply:
Yup I got a little confused here. I just couldn't find a way to get active lp from the vault api and tried to do this myself last minute so got things a little mixed up will resolve this